File: plugins/operators/prd_processor.py
"""
Operator to process a Product Requirements Document (PRD)
and extract domain entities for the DDD model
"""

from typing import Any, Dict, Optional, Sequence

from airflow.models import BaseOperator
from airflow.utils.decorators import apply_defaults

from plugins.hooks.supabase_hook import SupabaseHook
from plugins.utils.node_executor import NodeExecutor


class PRDProcessorOperator(BaseOperator):
    """
    Operator that processes a PRD artifact and extracts domain entities
    
    This operator takes a PRD artifact, processes it using a Node.js script,
    and creates a DDD model artifact with the extracted domain entities.
    """
    
    @apply_defaults
    def __init__(
        self,
        artifact_id: int,
        supabase_conn_id: str = 'supabase_default',
        *args, **kwargs
    ) -> None:
        """
        Initialize the PRD processor operator
        
        Args:
            artifact_id: The ID of the PRD artifact to process
            supabase_conn_id: The Supabase connection ID to use
        """
        super().__init__(*args, **kwargs)
        self.artifact_id = artifact_id
        self.supabase_conn_id = supabase_conn_id
    
    def execute(self, context: Dict[str, Any]) -> int:
        """
        Execute the PRD processing operation
        
        Args:
            context: The Airflow task context
            
        Returns:
            The ID of the created DDD model artifact
        """
        # Get the PRD artifact
        hook = SupabaseHook(self.supabase_conn_id)
        prd_artifact = hook.get_artifact(self.artifact_id)
        
        if not prd_artifact:
            raise ValueError(f"Artifact with ID {self.artifact_id} not found")
        
        if prd_artifact['type'] != 'PRD':
            raise ValueError(f"Artifact with ID {self.artifact_id} is not a PRD")
        
        # Process the PRD to extract domain entities
        self.log.info(f"Processing PRD artifact: {prd_artifact['name']}")
        
        # Execute the PRD processor
        result = NodeExecutor.execute_processor('prd_processor', {
            'prd': prd_artifact['content']
        })
        
        # Create a DDD model artifact
        ddd_model_artifact = {
            'name': f"DDD Model for {prd_artifact['name']}",
            'type': 'DDD_MODEL',
            'content': result,
            'createdBy': context['task_instance'].task_id,
            'parentId': prd_artifact['id'],
            'status': 'PENDING_APPROVAL'
        }
        
        # Save the DDD model artifact
        created_artifact = hook.create_artifact(ddd_model_artifact)
        
        self.log.info(f"Created DDD model artifact with ID: {created_artifact['id']}")
        
        # Get the next pipeline stage for approvals
        stage = hook.get_pipeline_stage_by_order(1)  # DDD Model is stage 1
        
        # Create approval tasks for stakeholders
        # In a real implementation, you would query stakeholders from the database
        # For simplicity, we're creating a single approval task here
        approval = {
            'artifactId': created_artifact['id'],
            'stakeholderId': 1,  # This should be dynamically determined
            'status': 'PENDING'
        }
        
        hook.create_approval(approval)
        
        return created_artifact['id']

File: plugins/operators/prd_processor.js
/**
 * Airflow operator for processing PRD documents and extracting domain entities
 */

const { extractDomainEntities, extractApiActions, createStructuredModel } = require('../../helpers/document_parser');
const SupabaseHook = require('../hooks/supabase_hook');

class PrdProcessorOperator {
  constructor(taskId, pipelineRunId, artifactId, stageId) {
    this.taskId = taskId;
    this.pipelineRunId = pipelineRunId;
    this.artifactId = artifactId;
    this.stageId = stageId;
    this.supabaseHook = new SupabaseHook();
  }

  /**
   * Execute the PRD processing task
   * @returns {Promise<Object>} - The execution result
   */
  async execute() {
    try {
      console.log(`Executing PRD processor task: ${this.taskId}`);
      
      // Create stage execution record
      const stageExecution = await this.supabaseHook.createStageExecution({
        pipelineRunId: this.pipelineRunId,
        stageId: this.stageId,
        artifactId: this.artifactId,
        status: 'running'
      });
      
      // Get the PRD artifact
      const prdArtifact = await this.supabaseHook.getArtifact(this.artifactId);
      if (!prdArtifact) {
        throw new Error(`PRD artifact with ID ${this.artifactId} not found`);
      }
      
      console.log(`Processing PRD: ${prdArtifact.title}`);
      
      // Extract domain entities and API actions
      const domainEntities = extractDomainEntities(prdArtifact.content);
      const apiActions = extractApiActions(prdArtifact.content);
      
      // Create structured model
      const structuredModel = createStructuredModel(domainEntities, apiActions);
      
      // Create new artifact for domain model
      const domainModelArtifact = await this.supabaseHook.createArtifact({
        title: `${prdArtifact.title} - Domain Model`,
        description: `Domain model extracted from ${prdArtifact.title}`,
        type: 'DDD',
        content: structuredModel,
        version: '1.0.0',
        parentId: prdArtifact.id,
        status: 'draft'
      });
      
      // Update stage execution to completed
      await this.supabaseHook.updateStageExecution(stageExecution.id, {
        status: 'completed',
        endTime: new Date(),
        artifactId: domainModelArtifact.id,
        logs: JSON.stringify({
          message: 'Successfully extracted domain model',
          entitiesCount: domainEntities.length,
          actionsCount: apiActions.length
        })
      });
      
      return {
        status: 'success',
        artifactId: domainModelArtifact.id,
        stageExecutionId: stageExecution.id
      };
    } catch (error) {
      console.error('Error in PRD processor:', error);
      
      // Update stage execution to failed
      if (this.stageExecution) {
        await this.supabaseHook.updateStageExecution(this.stageExecution.id, {
          status: 'failed',
          endTime: new Date(),
          logs: JSON.stringify({
            error: error.message,
            stack: error.stack
          })
        });
      }
      
      throw error;
    }
  }
}

module.exports = PrdProcessorOperator;
